src/fix_boundary_sources.cpp

```diff
@@ -0,0 +1,176 @@
+#include "config.h"
+
+#ifdef HAVE_MPI
+#ifdef NEED_UNDEF_SEEK_FOR_MPI
+// undef'ing SEEK_* is needed for MPICH, possibly other MPI versions
+#undef SEEK_SET
+#undef SEEK_END
+#undef SEEK_CUR
+#endif
+#include <mpi.h>
+#include <algorithm>
+#endif
+
+namespace meep {
+
+#ifdef HAVE_MPI
+static MPI_Comm mycomm = MPI_COMM_WORLD;
+#endif
+
+// data structure for sending source information from one chunk to another
+struct srcpt_info {
+  std::complex<double> A; // amplitude
+  ptrdiff_t index; // index in chunk's fields array
+  size_t src_time_id;
+  int chunk_idx;
+  int c; // component
+};
+
+// comparison function for sorting srcpt_info lexicographically
+// by (processor, src_time_id, chunk_idx, c)
+struct srcpt_info_compare {
+  fields_chunk **chunks;
+  bool operator() (srcpt_info a, srcpt_info b) {
+    int aproc = chunks[a.chunk_idx]->n_proc();
+    int bproc = chunks[b.chunk_idx]->n_proc();
+    return (aproc != bproc ? aproc < bproc :
+            (a.src_time_id != b.src_time_id ? a.src_time_id < b.src_time_id :
+             (a.chunk_idx != b.chunk_idx ? a.chunk_idx < b.chunk_idx :
+              a.c < b.c)));
+  }
+};
+
+void fields::fix_boundary_sources() {
+  am_now_working_on(Connecting);
+
+  std::vector<srcpt_info> boundarysources;
+
+  // find all not-owned source points and figure out in which chunk
+  // they are actually supposed to be located, storing info in boundarysources.
+  for (int i = 0; i < num_chunks; i++) {
+    FOR_FIELD_TYPES(ft) {
+      for (src_vol &src : chunks[i]->sources[ft])
+        if (src.needs_boundary_fix) {
+          for (size_t ipt = 0; ipt < src.num_points(); ++ipt) {
+            component c = src.c;
+            ivec here = chunks[i]->gv.iloc(c, src.index_at(ipt));
+            if (!chunks[i]->gv.owns(here) && src.amplitude(ipt) != 0.0) {
+              if (src.t()->id == 0) abort("bug: fix_boundary_sources called for non-registered source");
+
+              // find the chunk that owns this point, similar to logic in boundaries.cpp
+              std::complex<double> thephase;
+              if (locate_component_point(&c, &here, &thephase) && !on_metal_boundary(here)) {
+                for (int j = 0; j < num_chunks; j++)
+                  if (chunks[j]->gv.owns(here)) {
+                    srcpt_info s = { src.amplitude(ipt)*conj(thephase), chunks[j]->gv.index(c, here),  src.t()->id, chunks[j]->chunk_idx, c };
+                    boundarysources.push_back(s);
+                    break;
+                  }
+              }
+              src.set_amplitude(ipt, 0.0); // will no longer be needed
+            }
+          }
+          src.needs_boundary_fix = false;
+        }
+    }
+  }
+
+  // we need each process's data to be contiguous
+  srcpt_info_compare compare = {chunks};
+  std::sort(boundarysources.begin(), boundarysources.end(), compare);
+
+  // collect 2d (row-major) arrays offsets and numcomm,
+  // where numcomm[i,j] is the number of srcpt_info items
```

Comment:
Note: `numcomm[i,j]` here is stored at `numcomm[i*P + j]` (this is "row major order").